File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        self.complete = 0
        def isComplete(root):
            if not root:
                return
            else:
                self.complete+=1
                isComplete(root.left)
                isComplete(root.right)
        
        isComplete(root)
        return self.complete
        

File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if (!root) return 0;
        int levels = 1;
        int nodes = 0;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()) {
            int levelSize = q.size();
            if(levelSize < levels) {
                return nodes+levelSize;
            } else {
                levels<<=1;
                nodes+=levelSize;
            }
            for(int i=0; i<levelSize; ++i) {
                TreeNode* cur = q.front();
                q.pop();
                if(cur->left) q.push(cur->left);
                if(cur->right) q.push(cur->right);
            }
        }
        return nodes;
        
    }
};